src/Frontend/FrontendServiceProvider.php

```diff
@@ -83,17 +108,18 @@ public function addBaseCss(SourceCollector $sources)
     private function addLessVariables(SourceCollector $sources)
     {
         $sources->addString(function () {
+            $vars = $this->container->make('flarum.less.config');
             $settings = $this->container->make(SettingsRepositoryInterface::class);
 
-            $vars = [
-                'config-primary-color'   => $settings->get('theme_primary_color', '#000'),
-                'config-secondary-color' => $settings->get('theme_secondary_color', '#000'),
-                'config-dark-mode'       => $settings->get('theme_dark_mode') ? 'true' : 'false',
-                'config-colored-header'  => $settings->get('theme_colored_header') ? 'true' : 'false'
-            ];
+            return array_reduce(array_keys($vars), function ($string, $name) use ($vars, $settings) {
+                $var = $vars[$name];
+                $value = $settings->get($var['key'], $var['default'] ?? null);
+
+                if (isset($var['callback'])) {
+                    $value = $var['callback']($value);
+                }
 
-            return array_reduce(array_keys($vars), function ($string, $name) use ($vars) {
-                return $string."@$name: {$vars[$name]};";
+                return $string."@$name: {$value};";
```

Comment:
One of the things that worries me here is the possibility of injecting stuff dangerously (e.g. `0; @my-other-var: 1px`), but tbh that's not our problem.

It might be worth adding that injecting stuff here is possible, and so not to blatantly trust any user input.

Reply:
True, a bad input could result in the parser failing with a 500 error, possibly also locking the admin out if used in the admin styles as well. I'll adjust the docblock. Really need to look into having the parser gracefully fail instead, a badly compiled CSS is better than an error blocking access.

Reply:
> a bad input could result in the parser failing with a 500 error,

Could we throw https://github.com/flarum/core/blob/b8754c7d7db07bbea25d15e285da9dadeece72ce/src/Forum/ValidateCustomLess.php at it?

Reply:
Oh wow didn't know that class existed, it attempts a LESS compilation with the new settings, but only when the `custom_less` setting has been modified, we should use that instead of `RecompileFrontendAssets`.

Reply:
Yeah, we could configure that to respond to changes of all setting-linked variables, or build in a LESS-error-reporting system so build fails don't shut down the site.